src/systick.c

```diff
@@ -0,0 +1,82 @@
+#include "systick.h"
+#include "global_vars.h"
+#include "i2c.h"
+#include "adc.h"
+#include "output.h"
+#include "fan.h"
+#include "button.h"
+#include "led.h"
+#include "buzzer.h"
+
+#include <libopencm3/cm3/systick.h>
+#include <libopencm3/cm3/nvic.h>
+
+INA219_meas_t battery = {0};
+INA219_meas_t reg_5v = {0};
+int16_t board_temp = 0;
+
+uint8_t systick_slow_tick = 0;
+uint16_t systick_temp_tick = 0;
+uint8_t current_phase = 0;
+
+void systick_init(void) {
+    // Generate a 1ms systick interrupt
+    // 72MHz / 8 => 9000000 counts per second
+    systick_set_clocksource(STK_CSR_CLKSOURCE_AHB_DIV8);
+
+    // 9000000/9000 = 1000 overflows per second
+    // SysTick interrupt every N clock pulses
+    systick_set_reload(9000);
+
+    systick_interrupt_enable();
+
+    // Start counting.
+    systick_counter_enable();
+}
+
+void sys_tick_handler(void) {
```

Comment:
Feels odd that all this work is being done in an interrupt handler. Can non-time-critical things be moved to the main loop?

Reply:
This is by design, the timing of the actions inside the systick are deterministic and even when everything is run it totals  less than the tick period.

Notably the items that consume the majority of the time here (sampling INA219's and the ADCs) are the particularly time critical sections.